apis/core/v1beta1/function_webhook.go

```diff
@@ -38,4 +43,418 @@ var _ webhook.Defaulter = &Function{}
 // Default implements webhook.Defaulter so a webhook will be registered for the type
 func (r *Function) Default() {
 	functionlog.Info("default", "name", r.Name)
+	if r.Spec.Version == nil || *r.Spec.Version == "" {
+		version := "latest"
+		r.Spec.Version = &version
+	}
+}
+
+// +kubebuilder:webhook:path=/validate-core-openfunction-io-v1beta1-function,mutating=false,failurePolicy=fail,groups=core.openfunction.io,resources=functions,verbs=create;update,versions=v1beta1,name=vfunctions.of.io,sideEffects=None,admissionReviewVersions=v1
+var _ webhook.Validator = &Function{}
+
+func (r *Function) ValidateCreate() error {
+	functionlog.Info("validate create", "name", r.Name)
+	return r.Validate()
+}
+
+func (r *Function) ValidateUpdate(old runtime.Object) error {
+	functionlog.Info("validate update", "name", r.Name)
+	return r.Validate()
+}
+
+func (r *Function) ValidateDelete() error {
+	functionlog.Info("validate delete", "name", r.Name)
+	return nil
+}
+
+func (r *Function) Validate() error {
+	if r.Spec.Image == "" {
+		return errors.New("spec.image must be set")
+	}
+
+	if r.Spec.Build != nil {
+		if r.Spec.ImageCredentials == nil {
+			return errors.New("spec.imageCredentials must be set when `spec.build` is enabled")
+		}
+		if r.Spec.ImageCredentials != nil && r.Spec.ImageCredentials.Name == "" {
+			return errors.New("spec.imageCredentials.name must be set when `spec.build` is enabled")
+		}
+		if err := r.ValidateBuild(); err != nil {
+			return err
+		}
+	}
+
+	if r.Spec.Serving != nil {
+		if err := r.ValidateServing(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *Function) ValidateBuild() error {
+	if r.Spec.Build.Builder == nil && r.Spec.Build.Dockerfile == nil {
+		return errors.New("one of spec.build.builder and spec.build.dockerfile must exist")
+	}
+
+	if r.Spec.Build.SrcRepo != nil && r.Spec.Build.SrcRepo.Url == "" {
+		return errors.New("spec.build.srcRepo.url cannot be empty string")
+	}
+
+	if r.Spec.Build.Timeout != nil && r.Spec.Build.Timeout.Duration < 0 {
+		return errors.New("spec.build.timeout cannot be less than 0")
+	}
+
+	if r.Spec.Build.SuccessfulBuildsHistoryLimit != nil && *r.Spec.Build.SuccessfulBuildsHistoryLimit < 0 {
+		return errors.New("spec.build.successfulBuildsHistoryLimit cannot be less than 0")
+	}
+
+	if r.Spec.Build.FailedBuildsHistoryLimit != nil && *r.Spec.Build.FailedBuildsHistoryLimit < 0 {
```

Comment:
I think these can be achieved using annotations of [kubebuilder](https://book.kubebuilder.io/reference/markers/crd-validation.html).

Reply:
I think putting the validation all in the webhook makes it easier to unit test and avoid missing validation rules.

Reply:
Can we keep this part of the logic in the webhook instead of implementing it in the form of a kubebuilder annotation?

Reply:
Just a suggestion.